### PR TITLE
*: Fix column offset semantic

### PR DESCRIPTION
**Pull request by hicqu**
Writable columns means public columns + write-only columns.
Previously:
    update can touch writable columns;
    insert can touch public columns;
Now:
    both update and insert can touch writable columns.

### CODE REVIEW

File: executor/write.go [hunk 1]
```diff
@@ -854,23 +855,28 @@ func (e *InsertValues) getRowsSelect(cols []*table.Column) ([][]types.Datum, err
...
-	for i, v := range vals {
-		offset := cols[i].Offset
-		row[offset] = v
-		if !ignoreErr {
-			marked[offset] = struct{}{}
+	row := make([]types.Datum, len(e.Table.WritableCols()))
+	marked := make(map[int]struct{}, len(e.Table.WritableCols()))
+	for i, wc := range e.Table.WritableCols() {
+		// Here we must iterate on vals not cols, because vals maybe is empty.
+		for ii, v := range vals {
```
Comment (hicqu) [inline executor/write.go]: Here we use a nested loop, looks like ugly, but it can work.

File: plan/logical_plans.go [hunk 2]
```diff
@@ -264,6 +264,11 @@ type DataSource struct {
 
 	LimitCount *int64
 
+	// inUpdate indicates the plan is in update or not.
+	// If it's true, its schema will contain writable columns,
+	// otherwise it will contain public columns.
+	inUpdate bool
```
Comment (hanfei1991) [inline plan/logical_plans.go]: Why add this field? Will it be used in any other place?

File: executor/write.go (hunk 1 above)
Comment (hanfei1991) [inline executor/write.go]: use `j` instead of `ii`

File: plan/logical_plans.go (hunk 2 above)
Comment (hicqu) [inline plan/logical_plans.go]: No, but I think keep this field is a good idea, so we won't forget the DataSource schema contains public columns or writable columns.

**Pull request closed without merging**